File: server/skills/look_at_screen.js
// Skill: look at what's currently on screen and describe/evaluate it — without
// controlling anything. Deliberately separate from control_computer.js: this
// never starts a control session, never shows the red control overlay, never
// touches the mouse or keyboard, and needs no confirmation (looking is not an
// action with a side effect). It answers a single question about a single
// moment, not a repeating watch — for "tell me when X happens", see
// watch_for.js instead.
//
// Imports only control/ps-bridge.js (the same low-level screen/window reader
// control/session.js uses) and ai.js's askModel — never control/session.js
// itself, so this stays safe to import from anywhere skills/index.js reaches.

import { send as sendCommand } from '../control/ps-bridge.js';
import { saveScreenshot } from '../control/screenshot-store.js';
import { askModel } from '../ai.js';
import { broadcast } from '../events.js';
import { showIndicator, hideIndicator } from '../control/observation-bridge.js';

// Real bug found in live testing: the user is normally TALKING to Jarvis
// through a browser tab, so at the moment "look at my screen" runs, the OS
// foreground window (and the topmost window in z-order) is very often
// Jarvis's own tab — not whatever app the user actually means. The same
// ambiguity broke explicit targeting too: asking for "my Chrome window"
// while Jarvis also happens to be running in a Chrome tab could match either
// one, since both share the same process name. Fix: never treat Jarvis's own
// tab as a candidate unless it is genuinely the only window open — matched by
// a known browser process AND the page's real <title> (see public/index.html),
// not a bare "chrome" process check, so a File Explorer window titled
// "Jarvis-001" (a different process entirely) is never wrongly excluded.
const BROWSER_PROCESSES = new Set(['chrome', 'msedge', 'firefox', 'brave', 'opera', 'vivaldi']);
export function isJarvisOwnWindow(w) {
  if (!BROWSER_PROCESSES.has(String(w.processName || '').toLowerCase())) return false;
  const title = String(w.title || '').trim();
  return title === 'Jarvis' || /^Jarvis\s*[-–—]/.test(title);
}

/** Finds the best-matching open window for a free-text target, or null for "whichever is in front"/"the whole screen". Exported (alongside the default skill export) purely so it's directly testable per CLAUDE.md's pure-logic-module pattern — no server needed. */
export function pickWindow(windows, target) {
  // Exclude Jarvis's own tab from the candidate pool up front — only fall
  // back to including it if it's truly the only window open, since looking
  // at nothing at all would be a worse answer than looking at ourselves.
  const others = windows.filter((w) => !isJarvisOwnWindow(w));
  const pool = others.length ? others : windows;

  const cleaned = String(target || '').trim().toLowerCase();
  if (!cleaned || ['screen', 'desktop', 'everything', 'my screen', 'whole screen'].includes(cleaned)) {
    // windows[] is already topmost-first (EnumWindows' Z-order), so once
    // Jarvis's own tab is excluded, "the first one left" is a good proxy for
    // "whatever I was just looking at" even when nothing currently holds OS
    // input focus (e.g. focus is on Jarvis's text box, but the app the user
    // means is still sitting right behind it).
    return pool.find((w) => w.foreground) || pool[0] || null;
  }
  const byTitleOrProcess = pool.find(
    (w) => w.title.toLowerCase().includes(cleaned) || w.processName.toLowerCase().includes(cleaned)
  );
  return byTitleOrProcess || pool.find((w) => w.foreground) || pool[0] || null;
}

export default {
  name: 'look_at_screen',
  description:
    'Look at what is currently visible on the screen and describe or evaluate it, without taking any action or ' +
    'controlling anything. Use this for requests like "look at my screen and tell me what you think", "what does ' +
    'this say", or "is this window showing X" — never for a task where the user wants something actually done ' +
    '(use control_computer for that).',
  parameters: {
    type: 'object',
    properties: {
      question: {
        type: 'string',
        description: 'What the user wants to know or have evaluated about what\'s on screen. If they just said "look at my screen", describe generally.',
      },
      target: {
        type: 'string',
        description: 'Optional — which window to look at, by app name or part of its title (e.g. "chrome", "notepad"). Leave out to use whichever window is currently in front.',
      },
    },
    required: [],
  },
  async run({ question, target } = {}) {
    let windows = [];
    try {
      ({ windows } = await sendCommand('windows'));
    } catch {
      return { ok: false, error: "I couldn't reach the screen right now." };
    }

    const win = pickWindow(windows, target);
    if (!win && !windows.length) {
      return { ok: false, error: 'Nothing appears to be open on screen right now.' };
    }

    // The badge covers the whole "actually looking" span — from right before
    // the screenshot is captured through the model finishing its answer —
    // not just the instant sendCommand('screenshot') resolves, since the
    // point is to reflect real observation state, not just the moment the
    // OS-level pixel grab happens. Always released via finally, on every
    // return path below (success, model failure, or a thrown error).
    const observationToken = showIndicator('Jarvis is looking at your screen');
    try {
      let shot;
      try {
        shot = await sendCommand('screenshot', win ? { handle: win.handle } : {});
      } catch (err) {
        return { ok: false, error: err?.message || "I couldn't take a look at the screen." };
      }

      // Kept alongside control-session screenshots so there's one place to
      // review anything Jarvis has looked at, whether it acted or not.
      try {
        saveScreenshot(shot.base64, { label: win ? win.processName : 'screen' });
      } catch {
        // Not saving it is a lesser failure than not answering the question.
      }

      // A UIA text read is cheap and often more exact than a picture (real
      // button labels/values) — attached as extra context when available, and
      // it's also what lets a text-only model still answer instead of failing
      // outright on the vision requirement below.
      let elementsText = '';
      if (win) {
        try {
          const { elements } = await sendCommand('read_window', { handle: win.handle });
          if (elements?.length > 1) {
            elementsText = elements
              .slice(0, 100)
              .map((e) => `${e.role || 'element'}${e.name ? ` "${e.name}"` : ''}${e.value ? ` = "${String(e.value).slice(0, 200)}"` : ''}`)
              .join('\n');
          }
        } catch {
          // No readable tree — the screenshot alone is still useful.
        }
      }

      const askedQuestion = String(question || '').trim() || 'Describe what you see and share anything worth noting.';
      const windowNote = win ? `The window being looked at is "${win.title}" (${win.processName}).` : 'This is a capture of the whole screen — no single window was specified.';
      const promptText =
        `${windowNote}\n\n${elementsText ? `Its readable contents:\n${elementsText}\n\n` : ''}` +
        `The user asked: ${askedQuestion}\n\nAnswer plainly and conversationally, as if glancing over at their screen. Do not mention that you were given a screenshot or a list of elements — just answer.`;

      const result = await askModel({
        prompt: promptText,
        media: [{ kind: 'image', mimeType: shot.mimeType, dataBase64: shot.base64 }],
        need: { vision: true },
      });

      broadcast({ type: 'observed', target: win ? win.title : 'the screen' });

      if (!result.ok) {
        // A text-only fallback is only meaningful if we actually have real
        // window text to answer from — otherwise this is just failing quieter.
        if (elementsText) {
          const textOnly = await askModel({ prompt: promptText.replace(/\n\nThe user asked:/, '\n\n(No image is available for this model — answering from the window\'s text alone.)\n\nThe user asked:') });
          if (textOnly.ok) {
            return { ok: true, answer: textOnly.text, sawImage: false };
          }
        }
        return { ok: false, error: result.error };
      }

      return { ok: true, answer: result.text, sawImage: true };
    } finally {
      hideIndicator(observationToken);
    }
  },
};

File: backend/jarvis/tools/look_at_screen.py
"""Look at what is on screen and answer a question about it. Nothing else.

Deliberately separate from `control_computer`: this never starts a control
session, never shows the control bar, never touches the mouse or keyboard, and
needs no confirmation — looking is not an action with a side effect. It answers
one question about one moment; for "tell me when X happens" there is
`watch_for`.

Two things it does that a naive version would not:

* **It reads the window's own control tree as well as capturing it.** Real
  labels and values are more exact than a picture, and they are also what lets a
  text-only model answer at all instead of failing the vision requirement.
* **It never looks at Jarvis itself.** The user is normally talking to Jarvis
  through a browser tab, so the front window at the moment "look at my screen"
  runs is very often Jarvis's own — see `desktop.pick_window`.
"""

from __future__ import annotations

import base64

from ..capabilities import CapabilitySpec, Risk
from ..control import NoDesktopHere, describe_desktop, get_desktop, pick_window
from ..control.captures import SCREENSHOT, save

MAX_ELEMENT_LINES = 100


def _element_lines(elements: list) -> str:
    lines = []
    for element in elements[:MAX_ELEMENT_LINES]:
        piece = element.role or "element"
        if element.name:
            piece += f' "{element.name}"'
        if element.value:
            piece += f' = "{element.value[:200]}"'
        lines.append(piece)
    return "\n".join(lines)


def _run(question: str = "", target: str = "") -> dict:
    desktop = get_desktop()
    try:
        windows = desktop.windows()
    except NoDesktopHere as err:
        return {"ok": False, "error": str(err)}
    except Exception as err:  # noqa: BLE001
        return {"ok": False, "error": f"I couldn't reach the screen right now: {err}"}

    window = pick_window(windows, target)
    if window is None and not windows:
        return {"ok": False, "error": "Nothing appears to be open on screen right now."}

    from ..control.watching import observe

    with observe("Jarvis is looking at your screen"):
        try:
            shot = desktop.screenshot(window.handle if window else None)
        except Exception as err:  # noqa: BLE001
            return {"ok": False, "error": f"I couldn't take a look at the screen: {err}"}

        # Kept with the control session's own captures, so there is one place to
        # review anything Jarvis has looked at, whether it acted on it or not.
        try:
            save(SCREENSHOT, shot.data)
        except Exception:  # noqa: BLE001 — not saving it is a lesser failure than not answering
            pass

        elements_text = ""
        if window is not None:
            try:
                # Any real element counts. The original required more than one,
                # because its reader returned the window itself as an entry and
                # "just the window" meant nothing readable; this reader already
                # drops anything with neither a name nor a value, so a single
                # entry is a single piece of genuine content — often the text
                # area holding the whole document.
                elements = desktop.read_window(window.handle)
                if elements:
                    elements_text = _element_lines(elements)
            except Exception:  # noqa: BLE001 — the picture alone is still useful
                elements_text = ""

        asked = str(question or "").strip() or \
            "Describe what you see and share anything worth noting."
        where = (f'The window being looked at is "{window.title}" ({window.process_name}).'
                 if window else
                 "This is a capture of the whole screen — no single window was specified.")
        prompt = (f"{where}\n\n"
                  + (f"Its readable contents:\n{elements_text}\n\n" if elements_text else "")
                  + f"The user asked: {asked}\n\n"
                  "Answer plainly and conversationally, as if glancing over at their screen. "
                  "Do not mention that you were given a screenshot or a list of elements — "
                  "just answer.")

        from ..ai import ask_model

        reply = ask_model(
            prompt,
            media=[{"kind": "image", "mimeType": shot.mime_type,
                    "dataBase64": base64.b64encode(shot.data).decode("ascii")}],
            need={"vision": True})
        if reply.ok:
            return {"ok": True, "answer": reply.text, "sawImage": True}

        # Falling back to text is only meaningful when there IS real window text
        # to answer from; otherwise it is just failing more quietly.
        if elements_text:
            text_only = ask_model(prompt.replace(
                "\n\nThe user asked:",
                "\n\n(No image is available for this model — answering from the window's "
                "text alone.)\n\nThe user asked:"))
            if text_only.ok:
                return {"ok": True, "answer": text_only.text, "sawImage": False}
        return {"ok": False, "error": reply.error or "No model could look at that."}


SPEC = CapabilitySpec(
    id="builtin.look_at_screen",
    name="look_at_screen",
    description=("Look at what is currently visible on the screen and describe or evaluate "
                 "it, without taking any action or controlling anything. Use for \"look at my "
                 "screen and tell me what you think\", \"what does this say\", or \"is this "
                 "showing X\" — never for a task where something should actually be done "
                 "(use control_computer for that). " + describe_desktop()),
    input_schema={"type": "object", "properties": {
        "question": {"type": "string",
                     "description": "What they want to know about what's on screen. If they "
                                    "just said \"look at my screen\", describe generally."},
        "target": {"type": "string",
                   "description": "Optional — which window, by app name or part of its title. "
                                  "Leave out for whichever is in front."}},
        "required": []},
    # Looking changes nothing. The badge is what makes it visible, not a prompt.
    risk=Risk.LOW,
    handler=_run,
    timeout_s=90.0,
    tags=frozenset({"core"}),
)
